### PR TITLE
fast enumeration of all exits

### PR DESCRIPTION
**Pull request by karlthepagan**
Asking for Room.prototype.exits(number) will list room positions that represent each contiguous exit.

Usage:

Find one position representing all the separate portal structures in a room.
> `Game.rooms.E75S5.exits(0)`
< `[{"x":36,"y":19}]`

Find the single large top exit
> `Game.rooms.E75S5.exits(FIND_EXIT_TOP))`
< `[{"x":5,"y":0}]`

Find the single small right exit
> `Game.rooms.E75S5.exits(FIND_EXIT_RIGHT))`
< `[{"x":49,"y":35}]`

Find the three small bottom exits
> `JSON.stringify(Game.rooms.E75S5.exits(FIND_EXIT_BOTTOM))`
< `[{"x":7,"y":49},{"x":20,"y":49},{"x":32,"y":49}]`

Find the two small left exits
> `Game.rooms.E75S5.exits(FIND_EXIT_LEFT)`
 < `[{"x":0,"y":3},{"x":0,"y":24}]`

Find all the regular exits
> `Game.rooms.E75S5.exits(FIND_EXIT)`
< `[{"x":5,"y":0},{"x":7,"y":49},{"x":20,"y":49},{"x":32,"y":49},{"x":49,"y":35},{"x":0,"y":3},{"x":0,"y":24}]`

This isn't cached and `FIND_EXIT` doesn't return portals. I think both of those are worthwhile additions.

### CODE REVIEW

**Comment by karlthepagan** [comment]
To get the maximum point (furthest right/bottom) of an exit:

> `JSON.stringify(Game.rooms.E75S5.exits(FIND_EXIT_TOP, 1))`
< `[{"x":45,"y":0}]`

File: room.js [hunk 1]
```diff
@@ -1912,6 +1912,46 @@ mod.extend = function(){
...
+        let limit = -1;
+        const ret = [];
+        for (let i = 0; i < positions.length; i++) {
+            const pos = positions[i];
+            if (!(_.get(map,[pos.x-1, pos.y]) || _.get(map,[pos.x,pos.y-1]))) {
+                if (point && limit !== -1) {
+                    ret[limit].x += Math.ceil(point * (maxX - ret[limit].x));
+                    ret[limit].y += Math.ceil(point * (maxY - ret[limit].y));
+                }
+                limit++;
```
Comment (chobobobo) [inline room.js]: Is limit a running index of contiguous exits?

File: room.js [hunk 2]
```diff
@@ -1912,6 +1912,46 @@ mod.extend = function(){
...
+        }
+
+        // assuming in-order
+        let maxX, maxY;
+        let map = {};
+        let limit = -1;
+        const ret = [];
+        for (let i = 0; i < positions.length; i++) {
+            const pos = positions[i];
+            if (!(_.get(map,[pos.x-1, pos.y]) || _.get(map,[pos.x,pos.y-1]))) {
```
Comment (chobobobo) [inline room.js]: the previous position along the exit is not an exit?

Comment (karlthepagan) [inline room.js]: I consider it part of the same exit

File: room.js (hunk 1 above)
Comment (karlthepagan) [inline room.js]: Exactly that. By the end of this you will have an array of exits with the point relating to the percent along its top left to bottom right axis.

File: room.js (hunk 2 above)
Comment (karlthepagan) [inline room.js]: Oh, I guess you're puzzling out the logic - this if checks if the coordinates up and to the left is empty.

Comment (karlthepagan) [inline room.js]: !a &&!b ===!(a || b)

**Pull request merged**